BUG: Help python csv engine read binary buffers

fiendish opened the pull request:
The file buffer given to read_csv could have been opened in
binary mode, but the python csv reader errors on binary buffers.

- [x] closes #23779
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

fiendish committed:
BUG: Help python csv engine read binary buffers

The file buffer given to read_csv could have been opened in
binary mode, but the python csv reader errors on binary buffers.

closes #23779

fiendish committed:
whatsnew entry

fiendish commented:
What does Linux py37_np_dev know that I don't?

jreback commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -105,7 +105,7 @@ I/O
 
 - Avoid calling ``S3File.s3`` when reading parquet, as this was removed in s3fs version 0.3.0 (:issue:`27756`)
--
+- read_csv now accepts binary mode file buffers when using the Python csv engine (:issue:`23779`)
```
use ``:meth:`read_csv` ``; move this note to 1.0

jreback commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -296,3 +296,10 @@ def test_malformed_skipfooter(python_parser_only):
         parser.read_csv(StringIO(data), header=1, comment="#", skipfooter=1)
+
+
+def test_binary_buffer(python_parser_only, csv1):
```
run this for both csv engines

jreback commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -296,3 +296,10 @@ def test_malformed_skipfooter(python_parser_only):
+    # see gh-23779
+    parser = python_parser_only
+    with open(csv1, "rb") as f:
+        parser.read_csv(f)
```
assert the result contents

jreback replied:
move this test to parser/test_common.py see if you can co-locate with other buffer type tests.

fiendish committed:
satisfy gh-14418 for binary mode files

fiendish committed:
update tests per requested changes

test both csv engines,
assert equality between ascii and binary modes,
colocate with other buffer tests

fiendish committed:
move whatsnew note to 1.0

fiendish committed:
black formatting

fiendish commented:
Made the requested changes. Do you want a rebase to squash?

TomAugspurger commented on pandas/io/common.py:
```diff
@@ -342,9 +342,11 @@ def _get_handle(
         from s3fs import S3File
 
-        need_text_wrapping = (BytesIO, S3File)
+        need_text_wrapping = (BufferedIOBase, BytesIO, S3File)
```
BytesIO is a subclass of BufferedIOBytes. Can BytesIO be removed?

fiendish replied:
👍

TomAugspurger reviewed (commented):
No need to squash.

fiendish committed:
BytesIO is a subclass of BufferedIOBase

fiendish committed:
black formatting

fiendish commented:
Is it normal for random unrelated numpy errors to fail builds?

TomAugspurger commented:
Yes, we're debugging it still. Restarted azure.

jreback commented on pandas/io/common.py:
```diff
@@ -420,8 +422,10 @@ def _get_handle(
-        f = TextIOWrapper(f, encoding=encoding, newline="")
-        handles.append(f)
+        g = TextIOWrapper(f, encoding=encoding, newline="")
+        if not isinstance(f, no_close):
```
this is hard to follow, can you just use BufferedIOBase here rather than no_close?

jreback commented on pandas/tests/io/parser/test_common.py:
```diff
@@ -2020,9 +2020,32 @@ def test_file_handles_with_open(all_parsers, csv1):
+    parser = all_parsers
+
+    fpath = os.path.join(csv_dir_path, fname)
+
```
can you compare these versus the file read w/o open it?

something like this (for the 2 cases)

```
expected = parser.read_csv(fpath....)
result = .....
tm.assert_frame_equal(....)

result = ....
tm.assert....
```

fiendish commented:
> this is hard to follow, can you just use BufferedIOBase here rather than no_close?

I figured this way was more general in case there were future issues discovered beyond just BufferedIOBase, but consider it done.

fiendish committed:
remove no_close and compare with non-open read

fiendish committed:
Merge branch 'master' into patch-1

TomAugspurger commented:
Thanks @fiendish!
